experiments/quality_league.py

```diff
@@ -0,0 +1,477 @@
+                    match_up = list(reversed(match_up))
+
+                m = Match(args.partial_obs, match_up)
+                challenger = AI.get_or_none(name=m.p0)
+                defender = AI.get_or_none(name=m.p1)
+                
+                r = m.run(args.num_matches // 2)
+                for item in r:
+                    drawn = False
+                    if item == Outcome.WIN.value:
+                        winner = challenger
+                        loser = defender
+                    elif item == Outcome.DRAW.value:
+                        drawn = True
+                    else:
+                        winner = defender
+                        loser = challenger
+                        
+                    print(f"{winner.name} {'draws' if drawn else 'wins'} {loser.name}")
+                    
+                    winner_rating, loser_rating = rate_1vs1(
+                        Rating(winner.mu, winner.sigma),
+                        Rating(loser.mu, loser.sigma),
+                        drawn=drawn)
+
+                    winner.mu, winner.sigma = winner_rating.mu, winner_rating.sigma
+                    loser.mu, loser.sigma = loser_rating.mu, loser_rating.sigma
+                    winner.save()
+                    loser.save()
+                    
+                    MatchHistory(
+                        challenger=challenger,
+                        defender=defender,
+                        win=int(item == 1),
+                        draw=int(item == 0),
+                        loss=int(item == -1),
+                    ).save()
+        get_leaderboard().to_csv(f"{dbname}.csv", index=False)
+
+    # case 2: new AIs
+    else:
+        leaderboard = get_leaderboard_existing_ais(existing_ai_names)
+        new_ai_names = [ai_name for ai_name in args.evals if ai_name not in existing_ai_names]
+        for new_ai_name in new_ai_names:
+            ai = AI.get(name=new_ai_name)
+
+            while ai.sigma > args.highest_sigma:
+
+                match_qualities = []
+                for ai2_name in leaderboard["name"]:
+                    opponent_ai = AI.get(name=ai2_name)
+                    if ai.name == opponent_ai.name:
+                        continue
+                    match_qualities += [[opponent_ai, quality_1vs1(ai, opponent_ai)]]
+                
+                # sort by quality
+                match_qualities = sorted(match_qualities, key=lambda x: x[1], reverse=True)
+                print("match_qualities[:3]", match_qualities[:3])
+
+                # run a match if the quality of the opponent is high enough
+                top_3_ai = [item[0] for item in match_qualities[:3]]
+                opponent_ai = random.choice(top_3_ai)
+                match_up = (ai.name, opponent_ai.name)
+                match_quality = quality_1vs1(ai, opponent_ai)
+                print(f"the match up is ({ai}, {opponent_ai}), quality is {round(match_quality, 4)}")
```

Comment:
> "Quality" here stands for quality estimate done by TrueSkill, right?

No, so the implementation gets a `quality_score` based on Trueskill for each opponent (via the [`quality_1vs1(r1, r2)` API)](https://trueskill.org/#trueskill.quality_1vs1), sort them by `quality_score`, and randomly samples an opponent from the pool with the 3 highest `quality_score`.